lib/common.go

```diff
@@ -26,18 +29,139 @@ const (
 // the slave knows from which string permutation to start and at which string
 // permutation to stop.
 type Task struct {
-	ID        int
-	JobID     int
-	Algorithm string
-	Salt      string
-	Digest    string
-	CharSet   string
-	Length    int
-	Start     string
-	End       string
+	Job
+	JobID   int
+	ID      int
+	Start   []byte
+	TaskLen int
 }
 
 // ToJSON serializes a Task to JSON.
 func (t *Task) ToJSON() ([]byte, error) {
 	return json.Marshal(t)
 }
+
+// Algorithm is an enum for the supported key derivation functions
+type Algorithm int
+
+const (
+	PBKDF2 Algorithm = iota
+	BCRYPT
+	SCRYPT
+	ARGON2
+)
+
+// Alphabet
+type Alphabet int
+
+const (
+	Numerical Alphabet = iota
+	AlphaLower
+	AlphaMixed
+	AlphaNumLower
+	AlphaNumMixed
+)
+
+func (alph Alphabet) InitialCandidate(l int) []byte {
```

Comment:
If you use pointer receiver you can't write `AlphaLower.InitialCandidate`. That's why I went with value receiver.

From that reference:

If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense.